pallets/manta-pay/src/lib.rs

```diff
@@ -132,34 +142,33 @@ pub mod pallet {
 	#[pallet::hooks]
 	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {}
 
-	/// Shards of the merkle tree of UTXOs
+	/// UTXOs and Encrypted Notes Grouped by Shard
 	#[pallet::storage]
 	pub(super) type Shards<T: Config> =
-		StorageDoubleMap<_, Identity, u8, Identity, u64, ([u8; 32], EncryptedNote), ValueQuery>;
+		StorageDoubleMap<_, Identity, u8, Identity, u64, (Utxo, EncryptedNote), ValueQuery>;
 
-	/// Shard merkle trees
+	/// Shard Merkle Tree Paths
 	#[pallet::storage]
 	pub(super) type ShardTrees<T: Config> =
 		StorageMap<_, Identity, u8, UtxoMerkleTreePath, ValueQuery>;
 
-	/// Outputs of Utxo accumulator
+	/// Outputs of Utxo Accumulator
 	#[pallet::storage]
 	pub(super) type UtxoAccumulatorOutputs<T: Config> =
-		StorageMap<_, Identity, [u8; 32], (), ValueQuery>;
+		StorageMap<_, Identity, UtxoAccumulatorOutput, (), ValueQuery>;
 
-	/// Utxo set of MantaPay protocol
+	/// UTXO Set
 	#[pallet::storage]
-	pub(super) type UtxoSet<T: Config> = StorageMap<_, Identity, [u8; 32], (), ValueQuery>;
+	pub(super) type UtxoSet<T: Config> = StorageMap<_, Identity, Utxo, (), ValueQuery>;
 
-	/// Void number set
+	/// Void Number Set
 	#[pallet::storage]
-	pub(super) type VoidNumberSet<T: Config> = StorageMap<_, Identity, [u8; 32], (), ValueQuery>;
+	pub(super) type VoidNumberSet<T: Config> = StorageMap<_, Identity, VoidNumber, (), ValueQuery>;
 
-	/// Void number set insertion order
-	/// Each element of the key is an `u64` insertion order number of void number.
+	/// Void Number Ordered by Insertion
 	#[pallet::storage]
-	pub(super) type VoidNumberSetInsertionOrder<T: Config> =
-		StorageMap<_, Identity, u64, [u8; 32], ValueQuery>;
+	pub(super) type InsertionOrderedVoidNumbers<T: Config> =
+		StorageMap<_, Identity, u64, VoidNumber, ValueQuery>;
 
 	/// The size of Void Number Set
 	/// FIXME: this should be removed.
```

Comment:
VoidNumberSetSize is in use, why remove?

Reply:
That was a legacy issue: I thought the storage map itself can track how many element it stored. It turned out not the case.

Reply:
Substrate has a `CountedStorageMap` for that

https://github.com/paritytech/substrate/blob/7a09b82772b0597cdae224f4b124d660dda9e82b/frame/support/src/storage/types/counted_map.rs#L49

Reply:
Why resolve this without adfdressing it? Want to fix it later?